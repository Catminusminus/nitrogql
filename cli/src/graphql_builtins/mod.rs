use crate::ast::{
    base::{Ident, Keyword, Pos},
    r#type::{NamedType, NonNullType, Type},
    type_system::{
        ArgumentsDefinition, DirectiveDefinition, InputValueDefinition, ScalarTypeDefinition,
        TypeDefinition, TypeSystemDefinitionOrExtension,
    },
    value::{StringValue, Value},
};

/// Generate built-in definitions.
/// TODO: make this configurable
pub fn generate_builtins() -> Vec<TypeSystemDefinitionOrExtension<'static>> {
    let type_definitions = vec![
        scalar("Int"),
        scalar("Float"),
        scalar("String"),
        scalar("Boolean"),
        scalar("ID"),
    ];

    let directive_definitions = vec![
        directive(
            "skip",
            vec![(
                "if",
                Type::NonNull(Box::new(NonNullType {
                    r#type: Type::Named(NamedType {
                        name: ident("Boolean"),
                    }),
                })),
                None,
            )],
            vec!["FIELD", "FRAGMENT_SPREAD", "INLINE_FRAGMENT"],
        ),
        directive(
            "include",
            vec![(
                "if",
                Type::NonNull(Box::new(NonNullType {
                    r#type: Type::Named(NamedType {
                        name: ident("Boolean"),
                    }),
                })),
                None,
            )],
            vec!["FIELD", "FRAGMENT_SPREAD", "INLINE_FRAGMENT"],
        ),
        directive(
            "deprecated",
            vec![(
                "reason",
                Type::Named(NamedType {
                    name: ident("String"),
                }),
                Some(Value::StringValue(StringValue {
                    position: Pos::builtin(),
                    value: String::from("No longer supported"),
                })),
            )],
            vec![
                "FIELD_DEFINITION",
                "ARGUMENT_DEFINITION",
                "INPUT_FIELD_DEFINITION",
                "ENUM_VALUE",
            ],
        ),
        directive(
            "specifiedBy",
            vec![(
                "url",
                Type::NonNull(Box::new(NonNullType {
                    r#type: Type::Named(NamedType {
                        name: ident("String"),
                    }),
                })),
                None,
            )],
            vec!["SCALAR"],
        ),
    ];

    type_definitions
        .into_iter()
        .map(|def| TypeSystemDefinitionOrExtension::TypeDefinition(def))
        .chain(
            directive_definitions
                .into_iter()
                .map(|def| TypeSystemDefinitionOrExtension::DirectiveDefinition(def)),
        )
        .collect()
}

fn scalar(name: &str) -> TypeDefinition {
    TypeDefinition::Scalar(ScalarTypeDefinition {
        description: None,
        position: Pos::builtin(),
        name: ident(name),
        directives: vec![],
        scalar_keyword: keyword("scalar"),
    })
}

fn directive<'a>(
    name: &'a str,
    arguments: Vec<(&'a str, Type<'a>, Option<Value<'a>>)>,
    locations: Vec<&'a str>,
) -> DirectiveDefinition<'a> {
    DirectiveDefinition {
        description: None,
        position: Pos::builtin(),
        name: ident(name),
        arguments: if arguments.is_empty() {
            None
        } else {
            Some(ArgumentsDefinition {
                input_values: arguments
                    .into_iter()
                    .map(|(name, ty, default_value)| InputValueDefinition {
                        description: None,
                        position: Pos::builtin(),
                        name: ident(name),
                        r#type: ty,
                        default_value,
                        directives: vec![],
                    })
                    .collect(),
            })
        },
        repeatable: None,
        locations: locations.into_iter().map(ident).collect(),
        directive_keyword: keyword("directive"),
    }
}

fn ident(name: &str) -> Ident {
    Ident {
        name,
        position: Pos::builtin(),
    }
}

fn keyword(name: &str) -> Keyword {
    Keyword {
        name,
        position: Pos::builtin(),
    }
}
