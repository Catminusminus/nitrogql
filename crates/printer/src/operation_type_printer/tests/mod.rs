use insta::assert_snapshot;

use graphql_builtins::generate_builtins;
use nitrogql_ast::{OperationDocument, TypeSystemDocument};
use nitrogql_parser::{parse_operation_document, parse_type_system_document};
use nitrogql_semantics::ast_to_type_system;
use nitrogql_semantics::resolve_extensions;
use sourcemap_writer::JustWriter;

use crate::print_types_for_operation_document;
use crate::OperationTypePrinterOptions;

fn type_system() -> TypeSystemDocument<'static> {
    let mut doc = parse_type_system_document(
        "
            type User implements HasID {
                id: ID!
                name: String!
                type: UserType!
                age: Int
                posts: [HasID!]!
            }
            interface HasID {
                id: ID!
            }
            enum UserType {
                NormalUser
                PremiumUser
            }
            type Bot implements HasID {
                id: ID!
            }

            type Post implements HasID {
                id: ID!
                title: String!
                tags: [String!]
                body: String!
            }

            type Tweet implements HasId {
                id: ID!
                body: String!
            }

            input UserSearchQuery {
                age: Int
                name: String
                keywords: [String!]
            }

            type Query {
                me: User!
            }
            ",
    )
    .unwrap();
    doc.extend(generate_builtins());
    let doc = resolve_extensions(doc).unwrap();
    doc
}

#[test]
fn basic_type_printing() {
    let doc = parse_operation_document(
        "
        query {
            me {
                id name type age
            }
        }
        ",
    )
    .unwrap();
    let printed = print_document(&doc);
    assert_snapshot!(printed);
}

#[test]
fn fragment_spread() {
    let doc = parse_operation_document(
        "
        query {
            me {
                id name type age
                posts {
                    id
                    ...F
                }
            }
        }
        fragment F on HasID {
            id
        }
        ",
    )
    .unwrap();
    let printed = print_document(&doc);
    assert_snapshot!(printed);
}

#[test]
fn fragment_spread_cond() {
    let doc = parse_operation_document(
        "
        query {
            me {
                id name type age
                posts {
                    id
                    ...F
                    ...P
                }
            }
        }
        fragment F on HasID {
            id
        }
        fragment P on Post {
            title
            body
        }
        ",
    )
    .unwrap();
    let printed = print_document(&doc);
    assert_snapshot!(printed);
}

#[test]
fn fragment_inline_spread() {
    let doc = parse_operation_document(
        "
        query {
            me {
                id name type age
                posts {
                    ... {
                        id
                    }
                }
            }
        }
        ",
    )
    .unwrap();
    let printed = print_document(&doc);
    assert_snapshot!(printed);
}

#[test]
fn query_variables() {
    let doc = parse_operation_document(
        "
        query($foo: Int!, $bar: String) {
            me(foo: $foo, bar: $bar) {
                id name type age
            }
        }
        ",
    )
    .unwrap();
    let printed = print_document(&doc);
    assert_snapshot!(printed);
}

mod skip_include {
    use super::*;

    #[test]
    fn skipped_field_is_not_included() {
        let doc = parse_operation_document(
            "
            query {
                me {
                    id name type
                    age @skip(if: true)
                }
            }
            ",
        )
        .unwrap();
        let printed = print_document(&doc);
        assert_snapshot!(printed);
    }

    #[test]
    fn not_skipped_field_is_included() {
        let doc = parse_operation_document(
            "
            query {
                me {
                    id name type
                    age @skip(if: false)
                }
            }
            ",
        )
        .unwrap();
        let printed = print_document(&doc);
        assert_snapshot!(printed);
    }

    #[test]
    fn included_field_is_included() {
        let doc = parse_operation_document(
            "
            query {
                me {
                    id name type
                    age @include(if: true)
                }
            }
            ",
        )
        .unwrap();
        let printed = print_document(&doc);
        assert_snapshot!(printed);
    }

    #[test]
    fn not_included_field_is_not_included() {
        let doc = parse_operation_document(
            "
            query {
                me {
                    id name type
                    age @include(if: false)
                }
            }
            ",
        )
        .unwrap();
        let printed = print_document(&doc);
        assert_snapshot!(printed);
    }

    #[test]
    fn skip_variable() {
        let doc = parse_operation_document(
            "
            query($skip: Boolean!) {
                me {
                    id name type
                    age @skip(if: $skip)
                }
            }
            ",
        )
        .unwrap();
        let printed = print_document(&doc);
        assert_snapshot!(printed);
    }

    #[test]
    fn include_variable() {
        let doc = parse_operation_document(
            "
            query($include: Boolean!) {
                me {
                    id name type
                    age @include(if: $include)
                }
            }
            ",
        )
        .unwrap();
        let printed = print_document(&doc);
        assert_snapshot!(printed);
    }

    #[test]
    fn skip_and_include() {
        let doc = parse_operation_document(
            "
            query {
                me {
                    id @skip(if: false) @include(if: false)
                    name @skip(if: true) @include(if: false)
                    type @skip(if: false) @include(if: true)
                    age @skip(if: true) @include(if: true)
                }
            }
            ",
        )
        .unwrap();
        let printed = print_document(&doc);
        assert_snapshot!(printed);
    }

    #[test]
    fn emit_only_possible_cases() {
        let doc = parse_operation_document(
            "
            query($flag: Boolean!) {
                me {
                    id @skip(if: $flag)
                    name @include(if: $flag)
                }
            }
            ",
        )
        .unwrap();
        let printed = print_document(&doc);
        assert_snapshot!(printed);
    }
}

fn print_document(document: &OperationDocument) -> String {
    let mut result = String::new();
    let mut writer = JustWriter::new(&mut result);
    let schema = type_system();
    let schema = ast_to_type_system(&schema);
    print_types_for_operation_document(
        OperationTypePrinterOptions::default(),
        &schema,
        document,
        &mut writer,
    );
    result
}
