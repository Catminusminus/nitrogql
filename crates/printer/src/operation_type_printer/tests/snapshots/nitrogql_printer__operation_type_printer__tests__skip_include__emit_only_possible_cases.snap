---
source: crates/printer/src/operation_type_printer/tests/mod.rs
expression: printed
---
import type { TypedDocumentNode } from "@graphql-typed-document-node/core";
import type * as Schema from "";

type QueryResult = Schema.__SelectionSet<Schema.Query, {
  me: Schema.__SelectionSet<Schema.User, {
    id: Schema.ID;
  }, {}> | Schema.__SelectionSet<Schema.User, {
    name: Schema.String;
  }, {}>;
}, {}>;

type QueryVariables = {
  flag: Schema.Boolean;
};

declare const Query: TypedDocumentNode<QueryResult, QueryVariables>;

export { Query as default };


